src/System.Text.Primitives/System/Text/TextEncoding.cs

```diff
@@ -121,6 +123,156 @@ public override bool TryEncodeFromUtf16(ReadOnlySpan<char> utf16, Span<byte> buf
             return true;
         }
 
+        public override bool TryEncodeFromUnicode(ReadOnlySpan<UnicodeCodePoint> codePoints, Span<byte> buffer, out int bytesWritten)
+        {
+            var avaliableBytes = buffer.Length;
+            int bytesWrittenForCodePoint = 0;
+            bytesWritten = 0;
+
+            for (int i = 0; i < codePoints.Length; i++)
+            {
+                if (!TryEncodeFromUnicode(codePoints[i], buffer.Slice(bytesWritten), out bytesWrittenForCodePoint))
+                {
+                    bytesWritten = 0;
+                    return false;
+                }
+                bytesWritten += bytesWrittenForCodePoint;
+            }
+
+            return true;
+        }
+
+        public override bool TryEncodeFromUnicode(UnicodeCodePoint codePoint, Span<byte> buffer, out int bytesWritten)
+        {
+            var avaliableBytes = buffer.Length;
+            bytesWritten = 0;
+
+            var codepoint = (uint)codePoint;
+            if (codepoint <= 0x7f) // this if block just optimizes for ascii
+            {
+                if (bytesWritten + 1 > avaliableBytes)
+                {
+                    bytesWritten = 0;
+                    return false;
+                }
+                buffer[bytesWritten++] = (byte)codepoint;
+            }
+            else
+            {
+                Utf8EncodedCodePoint encoded;
+                encoded = new Utf8EncodedCodePoint(codePoint);
```

Comment:
Yeah, this seems wasteful. You encode the code point to byte fileds inside Utf8EncodedCodePoint (via Span), then you take the values out of the fileds one-by-one and write them to the span. You should rewrite the routine to directly write to span